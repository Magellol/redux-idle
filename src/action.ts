export const REDUX_IDLE_ACTION_TYPE = '@@REDUX_IDLE_IDLENESS_UPDATE';

export const createIdlenessUpdateAction = (isIdle: boolean) => ({
  isIdle,
  type: REDUX_IDLE_ACTION_TYPE,
});
