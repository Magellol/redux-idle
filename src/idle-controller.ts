import { throttle } from 'lodash-es';

export type Listener = (isIdle: boolean) => void;

const DEFAULT_EVENTS = [
  'mousemove',
  'mousedown',
  'keydown',
  'touchstart',
  'scroll',
];

export const idleControllerFactory = (options: {
  timeToWaitBeforeIdleness: number;
  events?: string[];
}) => {
  const { events = DEFAULT_EVENTS, timeToWaitBeforeIdleness } = options;
  let hasListenerAttached = false;
  let isUserIdle = false;

  const listenToEvents = (listener: EventListener) =>
    events.forEach(event => window.addEventListener(event, listener, true));

  const unsubscribeFromEvents = (listener: EventListener) =>
    events.forEach(event => window.removeEventListener(event, listener, true));

  return {
    subscribe(onChangeListener: Listener) {
      if (typeof onChangeListener !== 'function') {
        throw new Error('Expected "subscribe" to get a function as argument.');
      }

      if (hasListenerAttached) {
        throw new Error(
          'This idle controller has already one listener attached to it. Multiple listeners is not currently supported.',
        );
      }

      hasListenerAttached = true;
      const startUserIdleTimeout = window.setTimeout.bind(
        window,
        () => {
          isUserIdle = true;
          onChangeListener(isUserIdle);
        },
        timeToWaitBeforeIdleness,
      );

      let timeout = startUserIdleTimeout();

      const handleEvent: EventListener = throttle(() => {
        if (hasListenerAttached === false) {
          return;
        }

        if (isUserIdle) {
          isUserIdle = false;
          onChangeListener(isUserIdle);
        }

        clearTimeout(timeout);
        timeout = startUserIdleTimeout();
      }, 1000);

      listenToEvents(handleEvent);

      return () => {
        hasListenerAttached = false;
        unsubscribeFromEvents(handleEvent);
        clearTimeout(timeout);
      };
    },
  };
};
