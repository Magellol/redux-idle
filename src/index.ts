export { idleReducer } from './reducer';
export { idleMiddleware } from './middleware';
