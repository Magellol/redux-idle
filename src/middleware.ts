import { Middleware } from 'redux';
import { idleControllerFactory } from './idle-controller';

export type IdleMiddlewareFactory = (
  options: { timeToWaitBeforeIdleness: number },
) => Middleware;

export interface IdleUpdateAction {
  type: typeof REDUX_IDLE_ACTION_TYPE;
  isIdle: boolean;
}

export const REDUX_IDLE_ACTION_TYPE = 'REDUX_IDLE_STATUS_UPDATE';

export const idleMiddleware: IdleMiddlewareFactory = ({
  timeToWaitBeforeIdleness,
}) => {
  const controller = idleControllerFactory({
    timeToWaitBeforeIdleness,
  });

  return middlewareApi => {
    controller.subscribe(isIdle => {
      const action: IdleUpdateAction = { isIdle, type: REDUX_IDLE_ACTION_TYPE };
      middlewareApi.dispatch(action);
    });

    return next => action => next(action);
  };
};
