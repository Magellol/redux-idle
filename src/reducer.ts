import { Reducer } from 'redux';
import { REDUX_IDLE_ACTION_TYPE } from './middleware';

export type UserIdlenessReducerFactory = () => Reducer<boolean>;

export const userIdlenessReducer: UserIdlenessReducerFactory = () => {
  return (state = false, action) => {
    if (action.type === REDUX_IDLE_ACTION_TYPE) {
      return action.isIdle;
    }

    return state;
  };
};
