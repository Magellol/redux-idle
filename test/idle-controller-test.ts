import { idleControllerFactory } from '../src/idle-controller';

const noop = () => undefined;
const forceToBeIdle = () => jest.runAllTimers();

describe('idleControllerFactory', () => {
  beforeAll(() => jest.useFakeTimers());
  afterAll(() => jest.useRealTimers());

  afterEach(() => jest.clearAllTimers());

  test('idleController#subscribe should throw if it does not receive a function as argument.', () => {
    const controller = idleControllerFactory({
      timeToWaitBeforeIdleness: 0,
    });

    expect(() =>
      controller.subscribe(<any>'not a function'),
    ).toThrowErrorMatchingSnapshot();
  });

  test('idleController#subscribe should throw if we are trying to subscribe multiple times without cancelling.', () => {
    const controller = idleControllerFactory({
      timeToWaitBeforeIdleness: 0,
    });

    controller.subscribe(noop);
    expect(() => controller.subscribe(noop)).toThrowErrorMatchingSnapshot();
  });

  test('idleController#subscribe should notify its callback when the user goes idle.', () => {
    const callback = jest.fn();
    const controller = idleControllerFactory({
      timeToWaitBeforeIdleness: 0,
    });

    controller.subscribe(callback);
    forceToBeIdle();

    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenLastCalledWith(true);
  });

  test('idleController#subscribe should notify its callback when the user triggers activity after being idle.', () => {
    const fakeUserActivityEvent = new Event('mousemove');
    const callback = jest.fn();
    const controller = idleControllerFactory({
      timeToWaitBeforeIdleness: 0,
    });

    controller.subscribe(callback);
    forceToBeIdle();

    window.dispatchEvent(fakeUserActivityEvent);

    expect(callback).toHaveBeenCalledTimes(2);
    expect(callback.mock.calls[0][0]).toBe(true);
    expect(callback.mock.calls[1][0]).toBe(false);
  });

  test('idleController#subscribe should _not_ notify its callback when the user triggers activity without being idle first.', () => {
    const fakeUserActivityEvent = new Event('keydown');
    const callback = jest.fn();
    const controller = idleControllerFactory({
      timeToWaitBeforeIdleness: 0,
    });

    controller.subscribe(callback);
    window.dispatchEvent(fakeUserActivityEvent);

    expect(callback).toHaveBeenCalledTimes(0);
  });

  test('idleController#subscribe should return a function that unsubscribes the controller.', () => {
    const fakeUserActivityEvent = new Event('mousemove');
    const callback = jest.fn();
    const controller = idleControllerFactory({
      timeToWaitBeforeIdleness: 0,
    });

    const cancel = controller.subscribe(callback);
    cancel();

    forceToBeIdle();
    window.dispatchEvent(fakeUserActivityEvent);

    expect(callback).toHaveBeenCalledTimes(0);
  });

  test('idleController#subscribe should subscribe to custom passed in events.', () => {
    const fakeUserActivityEvent = new Event('customEvent');
    const callback = jest.fn();
    const controller = idleControllerFactory({
      timeToWaitBeforeIdleness: 0,
      events: ['customEvent'],
    });

    controller.subscribe(callback);
    forceToBeIdle();

    window.dispatchEvent(fakeUserActivityEvent);

    expect(callback).toHaveBeenCalledTimes(2);
  });
});
