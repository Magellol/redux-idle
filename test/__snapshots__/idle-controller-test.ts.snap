// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`idleControllerFactory idleController#subscribe should throw if it does not receive a function as argument. 1`] = `"Expected \\"subscribe\\" to get a function as argument."`;

exports[`idleControllerFactory idleController#subscribe should throw if we are trying to subscribe multiple times without cancelling. 1`] = `"This idle controller has already one listener attached to it. Multiple listeners is not currently supported."`;
